interfaces/builtin/docker.go

```diff
@@ -0,0 +1,185 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"github.com/snapcore/snapd/interfaces"
+)
+
+var dockerPermanentSlotAppArmor = []byte(`
+# Description: Allow operating as the Docker daemon. Reserved because this
+#  gives privileged access to the system.
+# Usage: reserved
+
+# Allow sockets
+/{,var/}run/docker.sock rw,
+/{,var/}run/docker/     rw,
+/{,var/}run/docker/**   mrwklix,
+/{,var/}run/runc/       rw,
+/{,var/}run/runc/**     mrwklix,
+
+# Wide read access to /proc, but somewhat limited writes for now
+@{PROC}/ r,
+@{PROC}/** r,
+@{PROC}/[0-9]*/attr/exec w,
+@{PROC}/sys/net/** w,
+@{PROC}/[0-9]*/cmdline r,
+@{PROC}/[0-9]*/oom_score_adj w,
+
+# Wide read access to /sys
+/sys/** r,
+# Limit cgroup writes a bit
+/sys/fs/cgroup/*/docker/   rw,
+/sys/fs/cgroup/*/docker/** rw,
+/sys/fs/cgroup/*/system.slice/   rw,
+/sys/fs/cgroup/*/system.slice/** rw,
+
+# Allow tracing other processes (especially ourselves)
+ptrace (read, trace),
+
+# Docker needs a lot of caps, but limits them in the app container
+capability,
+
+# Allow talking to systemd
+dbus (send)
+    bus=system
+    peer=(name=org.freedesktop.systemd*,label=unconfined),
+# Allow receiving from unconfined
+dbus (receive)
+    bus=system
+    peer=(label=unconfined),
+
+# Docker does all kinds of mounts all over the filesystem
+/dev/mapper/control rw,
+/dev/mapper/docker* rw,
+/dev/loop* r,
+/dev/loop[0-9]* w,
+mount,
+umount,
+pivot_root,
+/.pivot_root*/ rw,
+
+# for console access
+/dev/ptmx rw,
+# file descriptors (/proc/NNN/fd/X)
+/[0-9]*   rw,
+
+# Docker needs to be able to create and load the profile it applies to containers ("docker-default")
+# XXX We might be able to get rid of this if we generate and load docker-default ourselves and make docker not do it.
+/etc/apparmor.d/docker rw,
+/sbin/apparmor_parser ixr,
+/etc/apparmor*/** r,
+/etc/apparmor.d/cache/docker* w,
+/sys/kernel/security/apparmor/** rw,
+
+# We'll want to adjust this to support --security-opts...
+change_profile -> docker-default,
+signal (send) peer=docker-default,
+ptrace (read, trace) peer=docker-default,
+
+# Graph (storage) driver bits
+/dev/shm/aufs.xino rw,
+
+#cf bug 1502785
+/ r,
+`)
+
+var dockerConnectedPlugAppArmor = []byte(`
+# Description: Allow using Docker. Reserved because this gives
+#  privileged access to the service/host.
+# Usage: reserved
+
+# Obviously need to be able to talk to the daemon
+/{,var/}run/docker.sock rw,
+
+@{PROC}/sys/net/core/somaxconn r,
+`)
+
+var dockerPermanentSlotSecComp = []byte(`
+# The Docker daemon needs to be able to launch arbitrary processes within containers (whose syscall needs are unknown beforehand)
+@unrestricted
```

Comment:
To be slightly more specific regarding loading modules, here's the precise Docker daemon logs where it's using `modprobe`, getting `permission denied`, ignoring the error and hitting `iptables` anyhow, then failing:
```
time="2016-08-02T14:46:55.740368931-07:00" level=warning msg="Running modprobe bridge br_netfilter failed with message: , error: fork/exec /sbin/modprobe: permission denied"
time="2016-08-02T14:46:55.741108535-07:00" level=warning msg="Running modprobe nf_nat failed with message: ``, error: fork/exec /sbin/modprobe: permission denied"
time="2016-08-02T14:46:55.741794046-07:00" level=warning msg="Running modprobe xt_conntrack failed with message: ``, error: fork/exec /sbin/modprobe: permission denied"
time="2016-08-02T14:46:55.742125959-07:00" level=debug msg="Fail to initialize firewalld: Failed to connect to D-Bus system bus: dial unix /var/run/dbus/system_bus_socket: connect: permission denied, using raw iptables instead"
time="2016-08-02T14:46:55.743626077-07:00" level=debug msg="/sbin/iptables, [--version]"
time="2016-08-02T14:46:55.745027672-07:00" level=debug msg="/sbin/iptables, [-t nat -D PREROUTING -m addrtype --dst-type LOCAL -j DOCKER]"
time="2016-08-02T14:46:55.746875503-07:00" level=debug msg="/sbin/iptables, [-t nat -D OUTPUT -m addrtype --dst-type LOCAL ! --dst 127.0.0.0/8 -j DOCKER]"
time="2016-08-02T14:46:55.748674940-07:00" level=debug msg="/sbin/iptables, [-t nat -D OUTPUT -m addrtype --dst-type LOCAL -j DOCKER]"
time="2016-08-02T14:46:55.750387276-07:00" level=debug msg="/sbin/iptables, [-t nat -D PREROUTING]"
time="2016-08-02T14:46:55.751859031-07:00" level=debug msg="/sbin/iptables, [-t nat -D OUTPUT]"
time="2016-08-02T14:46:55.753300771-07:00" level=debug msg="/sbin/iptables, [-t nat -F DOCKER]"
time="2016-08-02T14:46:55.754844062-07:00" level=debug msg="/sbin/iptables, [-t nat -X DOCKER]"
time="2016-08-02T14:46:55.756263058-07:00" level=debug msg="/sbin/iptables, [-t filter -F DOCKER]"
time="2016-08-02T14:46:55.757613661-07:00" level=debug msg="/sbin/iptables, [-t filter -X DOCKER]"
time="2016-08-02T14:46:55.759091740-07:00" level=debug msg="/sbin/iptables, [-t filter -F DOCKER-ISOLATION]"
time="2016-08-02T14:46:55.760441146-07:00" level=debug msg="/sbin/iptables, [-t filter -X DOCKER-ISOLATION]"
time="2016-08-02T14:46:55.761857607-07:00" level=debug msg="/sbin/iptables, [-t nat -n -L DOCKER]"
time="2016-08-02T14:46:55.763431457-07:00" level=debug msg="/sbin/iptables, [-t nat -N DOCKER]"
time="2016-08-02T14:46:55.765514246-07:00" level=debug msg="Cleaning up old mountid : start."
time="2016-08-02T14:46:55.765758158-07:00" level=fatal msg="Error starting daemon: Error initializing network controller: error obtaining controller instance: failed to create NAT chain: iptables failed: iptables -t nat -N DOCKER: iptables v1.6.0: can't initialize iptables table `nat': Permission denied\nPerhaps iptables or your kernel needs to be upgraded.\n (exit status 3)"
```

(and this is with the `firewall-control` plug connected)